confmap/resolver.go

```diff
@@ -234,28 +241,38 @@ func (mr *Resolver) expandValueRecursively(ctx context.Context, value interface{
 func (mr *Resolver) expandValue(ctx context.Context, value interface{}) (interface{}, bool, error) {
 	switch v := value.(type) {
 	case string:
-		// If it doesn't have the format "${scheme:opaque}" no need to expand.
-		if !strings.HasPrefix(v, "${") || !strings.HasSuffix(v, "}") || !strings.Contains(v, ":") {
-			return value, false, nil
-		}
-		lURI, err := newLocation(v[2 : len(v)-1])
-		if err != nil {
-			// Cannot return error, since a case like "${HOST}:${PORT}" is invalid location,
-			// but is supported in the legacy implementation.
+		// If no embedded "uris" no need to expand. embeddedURI regexp matches uriRegexp as well.
+		if !embeddedURI.MatchString(v) {
 			return value, false, nil
 		}
-		if strings.Contains(lURI.opaqueValue, "$") {
-			return nil, false, fmt.Errorf("the uri %q contains unsupported characters ('$')", lURI.asString())
-		}
-		ret, err := mr.retrieveValue(ctx, lURI)
-		if err != nil {
-			return nil, false, err
+
+		// If the value is a single URI, then the return value can be anything.
+		// This is the case `foo: ${file:some_extra_config.yml}`.
+		if embeddedURI.FindString(v) == v {
+			return mr.expandStringURI(ctx, v)
 		}
-		mr.closers = append(mr.closers, ret.Close)
-		val, err := ret.AsRaw()
-		return val, true, err
-	case []interface{}:
-		nslice := make([]interface{}, 0, len(v))
+
+		// If the URI is embedded into the string, return value must be a string, and we have to concatenate all strings.
+		var nerr error
+		var nchanged bool
+		nv := embeddedURI.ReplaceAllStringFunc(v, func(s string) string {
+			ret, changed, err := mr.expandStringURI(ctx, s)
+			nchanged = nchanged || changed
+			nerr = multierr.Append(nerr, err)
+			if err != nil {
+				return ""
+			}
+			switch val := ret.(type) {
+			case string:
+				return val
+			default:
+				nerr = multierr.Append(nerr, fmt.Errorf("expanding %v, expected string value type, got %T", s, val))
+				return v
```

Comment:
Should we make the best effort to convert the value to string here? I'm not sure we can expand something like `${env:HOST}:${env:PORT}` this way because the port probably will be a integer

Reply:
That is not true, every env var is a string see https://pkg.go.dev/os#Getenv

Reply:
Getenv returns string, but we apply yaml.Unmarshal on it :) https://github.com/open-telemetry/opentelemetry-collector/blob/main/confmap/provider/internal/provider.go#L28

Reply:
We should probably do the same even if it's just wrapped into the string quotes`"${env:PORT}"`

Reply:
Done. Added support to all supported values in retrieve as well.